torch/jit/__init__.py

```diff
@@ -631,10 +634,31 @@ def __getattr__(self, attr):
         return self.module._get_method(attr)
 
 
-def script(fn, optimize=True, _frames_up=0):
+def weak_script(fn, _frames_up=0):
+    compiled_weak_fns[fn] = {
+        "status": COMPILATION_PENDING,
+        "compiled_fn": None,
+        "rcb": createResolutionCallback(_frames_up + 1)
+    }
+    return fn
+
+
+def _try_compile_weak_script(fn):
+    entry = compiled_weak_fns.get(fn)
+    if entry is None:
+        return None
+    if entry["status"] == COMPILATION_PENDING:
+        entry["status"] = COMPILED
```

Comment:
This should probably be set only as the last line of this block (for thread safety)